pkg/util/limiter/utilization.go

```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package limiter
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+	"github.com/pkg/errors"
+	"github.com/prometheus/procfs"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/mimir/pkg/util/math"
+)
+
+const (
+	// Interval for updating resource (CPU/memory) utilization
+	resourceUtilizationUpdateInterval = time.Second
+
+	memPathV2 = "/sys/fs/cgroup/memory.current"
+	cpuPathV2 = "/sys/fs/cgroup/cpu.stat"
+	memPathV1 = "/sys/fs/cgroup/memory/memory.stat"
+	cpuPathV1 = "/sys/fs/cgroup/cpu/cpuacct.usage"
+)
+
+type utilizationScanner interface {
+	fmt.Stringer
+
+	// Scan returns CPU time in seconds and memory utilization in bytes, or an error.
+	Scan() (float64, uint64, error)
+}
+
+type cgroupV2Scanner struct {
+}
+
+func (s cgroupV2Scanner) String() string {
+	return "cgroup v2"
+}
+
+func (s cgroupV2Scanner) Scan() (float64, uint64, error) {
```

Comment:
I would like to see a unit test for this function. To make it easy to test you can move all the logic (except the call to `readFileNoStat()` to a new function `scan()` which takes in input the output of `readFileNoStat()` so that you can control the reader passed to `scan()` in the unit test.

Reply:
As we discussed elsewhere, I've removed the cgroup logic.